### PR TITLE
1050: Fix PCIeTopology ResetLink PATCH Handling

### PR DESCRIPTION
**Pull request by baemyung**
In PATCH logic, readJson() is incorrectly used and it causes the error

```
curl -k -X PATCH https://$bmc/redfish/v1/Systems/system/PCIeDevices/chassis_motherboard_pcieslot8_pcie_card8 \
     -d '{"Oem": {"IBM": {"LinkReset": true}}}'
{
  "Oem@Message.ExtendedInfo": [
    {
      "@odata.type": "#Message.v1_1_1.Message",
      "Message": "The value 'null' for the property Oem is of a different type than the property can accept.",
      "MessageArgs": [
        "null",
        "Oem"
      ],
      "MessageId": "Base.1.13.0.PropertyValueTypeError",
      "MessageSeverity": "Warning",
      "Resolution": "Correct the value for the property in the request body and resubmit the request if the operation failed."
    }
  ]
}
```

It is used to use the proper json data type

```
        std::optional<std::vector<nlohmann::json>> slotsData;
-->
        std::optional<nlohmann::json> slotsData;
        if (!json_util::readJsonPatch(req, asyncResp->res, "Oem", slotsData))
        {
            return;
        }
```

Tested:
 - Redfish Validator passed
 - Execute LinkReset PATCH

### CODE REVIEW

File: redfish-core/lib/pcie.hpp [hunk 1]
```diff
@@ -712,7 +712,7 @@ inline void requestRoutesSystemPCIeDevice(App& app)
             [](const crow::Request& req,
                const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
                const std::string& device) {
-        std::optional<std::vector<nlohmann::json>> slotsData;
+        std::optional<nlohmann::json> slotsData;
```
Comment (gtmills) [inline redfish-core/lib/pcie.hpp:715]: would be good to call this oem / oemData / something other than slotsData. E.g. https://github.com/ibm-openbmc/bmcweb/blob/07faf31c26909fb3cdcfc77f82c682bd58b50cb3/redfish-core/lib/managers.hpp#L2315

Comment (baemyung) [reply redfish-core/lib/pcie.hpp]: > would be good to call this oem / oemData / something other than slotsData. E.g.
> 
> https://github.com/ibm-openbmc/bmcweb/blob/07faf31c26909fb3cdcfc77f82c682bd58b50cb3/redfish-core/lib/managers.hpp#L2315

Changed.

File: redfish-core/lib/pcie.hpp [hunk 2]
```diff
@@ -712,40 +712,34 @@ inline void requestRoutesSystemPCIeDevice(App& app)
...
                const std::string& device) {
-        std::optional<std::vector<nlohmann::json>> slotsData;
-        if (!json_util::readJsonPatch(req, asyncResp->res, "Oem", slotsData))
+        std::optional<nlohmann::json> oem;
+        if (!json_util::readJsonPatch(req, asyncResp->res, "Oem", oem))
         {
             return;
         }
-        if (!slotsData || slotsData->empty())
+        if (!oem || oem->empty())
```
Comment (gtmills) [inline redfish-core/lib/pcie.hpp:720]: The  oem->empty() made sense when slotsData was a vector but I don't think here it is needed

Comment (baemyung) [reply redfish-core/lib/pcie.hpp:720]: Those invalid requests give 400.

For example,
```
redfishtool raw PATCH -r ${bmc}  -u service -p 0penBmc0 -S Always redfishtool raw PATCH -r ${bmc}  -u service -p 0penBmc0 -S Always /redfish/v1/Systems/system/PCIeDevices/chassis_motherboard_pcieslot8_pcie_card8  -d '{"Oem": {"IBM": {}}}'
   redfishtool: Transport: Response Error: status_code: 400 -- Bad Request
...
redfishtool raw PATCH -r ${bmc}  -u service -p 0penBmc0 -S Always /redfish/v1/Systems/system/PCIeDevices/chassis_motherboard_pcieslot8_pcie_card8  -d '{"Oem": {}}'
   redfishtool: Transport: Response Error: status_code: 400 -- Bad Request
```
I'll check out whether the checks like `if(!ibmOem)` is needed.

Comment (baemyung) [reply redfish-core/lib/pcie.hpp:720]: So I fixed the code by adding those check and return error msg like

```
+        if (!ibmOem || ibmOem->empty())
+        {
+            messages::propertyMissing(asyncResp->res, "IBM");
+            return;
+        }
+
```

**Review by gtmills — APPROVED** [review]

File: redfish-core/lib/pcie.hpp [hunk 3]
```diff
@@ -712,40 +712,45 @@ inline void requestRoutesSystemPCIeDevice(App& app)
...
             return;
         }
 
-        std::vector<nlohmann::json> slots = std::move(*slotsData);
+        std::optional<nlohmann::json> ibmOem;
+        if (!json_util::readJson(*oem, asyncResp->res, "IBM", ibmOem))
+        {
+            return;
+        }
+        if (!ibmOem || ibmOem->empty())
```
[the remark this answers is not in the document]
Comment (baemyung) [reply redfish-core/lib/pcie.hpp]: Thanks for tips.  I changed it.

**Pull request merged**